### PR TITLE
[CARBONDATA-1318]Fixed Concurrent table data loading unsafe memory issue

### PR DESCRIPTION
**Pull request by kumarvishal09**
Fixed task cancellation leak issue
Fixed task cleanup issue in data loading
Fixed Concurrent table data loading unsafe memory issue

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3117/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/526/

File: core/src/main/java/org/apache/carbondata/core/memory/UnsafeMemoryManager.java [hunk 1]
```diff
@@ -75,76 +79,67 @@
...
-  private synchronized MemoryBlock allocateMemory(long memoryRequested) {
+  private synchronized MemoryBlock allocateMemory(long taskId, long memoryRequested) {
     if (memoryUsed + memoryRequested <= totalMemory) {
       MemoryBlock allocate = allocator.allocate(memoryRequested);
       memoryUsed += allocate.size();
-      if (LOGGER.isDebugEnabled()) {
-        set.add(allocate);
-        LOGGER.error("Memory block (" + allocate + ") is created with size "  + allocate.size() +
-            ". Total memory used " + memoryUsed + "Bytes, left " + getAvailableMemory() + "Bytes");
+      Set<MemoryBlock> listOfMemoryBlock = taskIdToMemoryBlockMap.get(taskId);
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/memory/UnsafeMemoryManager.java]: Better add log to know how much is allocated so far and how much is left

File: core/src/main/java/org/apache/carbondata/core/memory/UnsafeMemoryManager.java [hunk 2]
```diff
@@ -75,76 +79,67 @@
...
       }
+      listOfMemoryBlock.add(allocate);
       return allocate;
     }
     return null;
   }
 
-  public synchronized void freeMemory(MemoryBlock memoryBlock) {
+  public synchronized void freeMemory(long taskId,MemoryBlock memoryBlock) {
+    taskIdToMemoryBlockMap.get(taskId).remove(memoryBlock);
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/memory/UnsafeMemoryManager.java]: Please add log

File: core/src/main/java/org/apache/carbondata/core/memory/UnsafeSortMemoryManager.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  /**
+   * Get total memory available
+   *
+   * @return amount of memory available
+   */
+  public long getUsableMemory() {
+    return totalMemory;
+  }
+
+  public synchronized void freeMemory(long taskId, MemoryBlock memoryBlock) {
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/memory/UnsafeSortMemoryManager.java]: Please add log

Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/memory/UnsafeSortMemoryManager.java]: ok

File: core/src/main/java/org/apache/carbondata/core/memory/UnsafeMemoryManager.java (hunk 2 above)
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/memory/UnsafeMemoryManager.java]: ok

File: core/src/main/java/org/apache/carbondata/core/memory/UnsafeMemoryManager.java (hunk 1 above)
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/memory/UnsafeMemoryManager.java]: ok

Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/memory/UnsafeMemoryManager.java]: @ravipesala Debug Log or info log??

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/530/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3121/

File: core/src/main/java/org/apache/carbondata/core/memory/UnsafeSortMemoryManager.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    if (null != memoryBlocks) {
+      Iterator<MemoryBlock> iterator = memoryBlocks.iterator();
+      while (iterator.hasNext()) {
+        allocator.free(iterator.next());
+      }
+    }
+  }
+
+  public synchronized MemoryBlock allocateMemoryLazy(long taskId, long memoryRequested) {
+    MemoryBlock allocate = allocator.allocate(memoryRequested);
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/memory/UnsafeSortMemoryManager.java]: add log

File: core/src/main/java/org/apache/carbondata/core/memory/UnsafeSortMemoryManager.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        } catch (InterruptedException e) {
+          throw new MemoryException(e);
+        }
+      } else {
+        break;
+      }
+      tries++;
+    }
+    if (baseBlock == null) {
+      throw new MemoryException("Not enough memory");
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/memory/UnsafeSortMemoryManager.java]: Ideally this case should not come in any case

File: core/src/main/java/org/apache/carbondata/core/memory/UnsafeSortMemoryManager.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    }
+    if (null != memoryBlocks) {
+      Iterator<MemoryBlock> iterator = memoryBlocks.iterator();
+      while (iterator.hasNext()) {
+        allocator.free(iterator.next());
+      }
+    }
+  }
+
+  public synchronized MemoryBlock allocateMemoryLazy(long taskId, long memoryRequested) {
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/memory/UnsafeSortMemoryManager.java]: Can you add comment what is it used for

File: core/src/main/java/org/apache/carbondata/core/memory/UnsafeSortMemoryManager.java [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+      tries++;
+    }
+    if (baseBlock == null) {
+      throw new MemoryException("Not enough memory");
+    }
+    return baseBlock;
+  }
+
+  private synchronized MemoryBlock allocateMemory(long taskId, long memoryRequested) {
+    if (memoryUsed + memoryRequested <= totalMemory) {
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/memory/UnsafeSortMemoryManager.java]: add log

Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/memory/UnsafeSortMemoryManager.java]: ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3127/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/536/

File: core/src/main/java/org/apache/carbondata/core/memory/IntPointerBuffer.java [hunk 8]
```diff
@@ -67,24 +73,25 @@ public int get(int rowId) {
...
+      pointerMemoryBlock =
+          UnsafeSortMemoryManager.allocateMemoryWithRetry(this.taskId, pointerBlock.length * 4);
+      for (int i = 0; i < pointerBlock.length; i++) {
+        CarbonUnsafe.unsafe
+            .putInt(pointerMemoryBlock.getBaseObject(), pointerMemoryBlock.getBaseOffset() + i * 4,
+                pointerBlock[i]);
+      }
+      pointerBlock = null;
+    } catch (MemoryException e) {
+      LOGGER.error("Not enough memory for allocating pointer buffer, sorting in heap");
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/memory/IntPointerBuffer.java]: change to `warn`

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/NewCarbonDataLoadRDD.scala [hunk 9]
```diff
@@ -454,6 +467,12 @@ class NewDataFrameLoaderRDD[K, V](
             .printStatisticsInfo(model.getPartitionId)
         }
       }
+      def clearMemory() {
```
Comment (kumarvishal09) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/NewCarbonDataLoadRDD.scala]: ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3152/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/559/

File: core/src/main/java/org/apache/carbondata/core/memory/UnsafeSortMemoryManager.java (hunk 5 above)
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/memory/UnsafeSortMemoryManager.java]: In caseof concurrent scenario this issue may come.

**Comment by ravipesala** [comment]
LGTM

**Pull request closed without merging**